### PR TITLE
Understanding the Flutter Getx ecosystem for state management

### LINKED ISSUE #3259: Understanding the Flutter Getx ecosystem for state management.
### Proposed title of article
Understanding the Flutter Getx ecosystem for state management.

### Introduction paragraph (2-3 paragraphs):
State management enables the sharing of application state between screens, across your app. Put differently, it allows you to share data from one UI to another so that when the state of your app changes, you change the state, and that triggers a redraw of the user interface. Flutter traditionally uses the Stateful Widget to manage state. While this works, but as your app grows bigger with nested widgets, it becomes a little difficult to pass data from a parent widget to a child widget and vice versa. In this kind of situation, you will have to pass the data through the constructors of the children widgets. The side effect is that data are passed to widgets that don't need them, just because they are within the widget tree. Also, the business logic will be tightly coupled to the User interface.

The GetX state management package solves this problem and many more. GetX is an extra-light and powerful solution for Flutter. It combines high-performance state management, intelligent dependency injection, and route management quickly and practically. It helps Developers achieve a high level of productivity through easy and pleasant syntax without sacrificing the maximum performance your application can deliver. It allows for total decoupling of the View, presentation logic, business logic, dependency injection, and navigation. This helps to achieve clean code by default.

 In this tutorial, we will build a simple Flutter mobile application to demonstrate the power of the GetX package.

### Key takeaways: 
- How to setup Flutter project and configure  dependencies
- How to use GetX as a state management tool
- Using Obx to maximize the power of reactive programming
- Exploring the Navigator Manager.
- How to use the GetBuilder sub ecosystem to manage state.
- Reusable component of the GetX ecosystem.
- Combining two approaches using MixinBuilder

### PR DESCRIPTION
**Pull request by Lekwacious**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes

This closes #3259.

### CODE REVIEW

**Comment by louisefindlay23** [comment]
@Lekwacious, I have copied and pasted the article checklist back into your PR comment. In the future, please do not delete this. Now you should go through the checklist and check the steps as you have completed them so your article is ready for review.

**Comment by Lekwacious** [comment]
Done. Thank you @louisefindlay23.

**Comment by Lekwacious** [comment]
Hi @louisefindlay23 no body has been assigned to review my PR. Just checking on the status of the PR

**Comment by louisefindlay23** [comment]
@WanjaMIKE, if you're free, could you take a look at this article? I don't have experience with Flutter but I know that you have.

**Comment by WanjaMIKE** [comment]
@louisefindlay23 Okay, I will.

File: content/articles/understanding-the-flutter-getx-ecosystem-for-state-management/index.md [hunk 1]
```diff
@@ -0,0 +1,983 @@
+
+
+### Introduction
+State management allows for the sharing of application state between screens, across your app. Put differently, it enables you to pass data from one User Interface to another so that when the state of your app changes, you change the state, and that rebuilds the user interface. Flutter traditionally uses the Stateful Widget to manage the state. While this works, but as your app grows bigger with nested widgets, it becomes a little difficult to pass data from a parent widget to a child widget and vice versa. In this kind of situation, you will have to pass the data through the constructors of the child widgets. The side effect is that data are passed to widgets that don't need them, just because they are within the widget tree. Also, the business logic will be tightly coupled to the User interface.
```
Comment (WanjaMIKE) [inline content/articles/understanding-the-flutter-getx-ecosystem-for-state-management/index.md:4]: Break into 2-3 paragraphs

File: content/articles/understanding-the-flutter-getx-ecosystem-for-state-management/index.md [hunk 2]
```diff
@@ -0,0 +1,983 @@
...
+- Give the project a name. Let’s call it ``shopping_getx``.
+- Choose a location where you want to save your project.
+- Then a project description.
+- Let’s name the organization com.lekwacious.
+- Leave the default Andriod and ios languages to Kotlin and Swift respectively.
+- Select the platform you want to develop for. Let's select and build for Andriod and ios. If you want to build for the web feel free to select the web option too then click on the finish button. ![Adding project details](project-details.png) 
+This will generate the default Flutter counter project in the main.dart file.
+
+### Integrating the GetX Ecosystem
+![GetX Documentation](getx-documentation.png)
```
Comment (WanjaMIKE) [inline content/articles/understanding-the-flutter-getx-ecosystem-for-state-management/index.md:38]: What is GetX. Could explain more on it

File: content/articles/understanding-the-flutter-getx-ecosystem-for-state-management/index.md [hunk 3]
```diff
@@ -0,0 +1,983 @@
...
+- What we need in the controller is the element we want to bind to the User Interface. In this case a list of ``Products``, as such we have created the field ``_items`` which contains the list of all the products. The underscore in the ``_items`` makes it private. We are hard-coding this, we can equally retrieve the product from the backend by making an HTTP call to the backend server.
+- Next, we created two getters that return all the list of the product and the product that has been marked as favorite respectively.
+- The ``findProductById`` method takes in an ``ID`` as an argument and returns the product with that particular ``ID``.
+- The ``toggleFavouriteStatus`` method takes in an ``ID`` and marks the product with that ``ID`` as favorite.
+- ``Update``: We called the update from ``Getx`` to update the User Interface on-click. The update method listens for changes in the ``toggleFavouriteStatus`` method and updates the appropriate User Interface. If you are familiar with the Provider package, the ``update`` method functions like the ``notifyListeners`` of the Provider package.
+
+#### The Cart Controller
```
Comment (WanjaMIKE) [inline content/articles/understanding-the-flutter-getx-ecosystem-for-state-management/index.md:264]: What will the cart controller do? Explain

File: content/articles/understanding-the-flutter-getx-ecosystem-for-state-management/index.md [hunk 4]
```diff
@@ -0,0 +1,983 @@
...
+        Order(
+            id: DateTime.now().toString(),
+            products: cartProducts,
+            amount: total,
+            dateTime: DateTime.now()));
+  }
+}
+ ```
+ 
+- ``obs``: We have declared a variable that is going to hold a list of all orders and we made it an observable by making it an obs using the dot notation. By making it an obs, every time the orders change, all widgets that use it are automatically changed.
```
Comment (WanjaMIKE) [inline content/articles/understanding-the-flutter-getx-ecosystem-for-state-management/index.md:967]: use ``` `obs` ```

**Review by WanjaMIKE — CHANGES_REQUESTED** [review]
Please heck the following comments

**Comment by ahmadmardeni1** [comment]
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept. @Lekwacious 
Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.
Be sure to see our resources page to see more info on plagiarism and what is considered as such.
cc @WanjaMIKE @louisefindlay23 

[Understanding the Flutter Getx ecosystem for state management#3561](https://github.com/section-engineering-education/engineering-education/files/7153709/New.Scan.10_57.AM.Copyleaks.Report.pdf)

**Comment by Lekwacious** [comment]
@WanjaMIKE  Thanks for the review. I have effected the changes you requsted

**Comment by WanjaMIKE** [comment]
@Lekwacious did you correct the plagiarism?

**Comment by Lekwacious** [comment]
@WanjaMIKE  Yes I did correct the plagiarism

**Comment by Kayere** [comment]
Great article @Lekwacious, keep up the good work 🚀.

**Pull request merged**